neps/nep-0522.md

```diff
@@ -0,0 +1,203 @@
+---
+NEP: 522
+Title: Stateless Transactions
+Authors: David Millar-Durrant <david@near.org>
+Status: Draft
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/522
+Type: Protocol
+Version: 1.0.0
+Created: 2023-12-05
+LastUpdated: 2023-12-05
+---
+
+## Summary
+
+This allows the sending of transactions to the NEAR Blockchain without depending on any Blockchain state. It also allows individual keys to send transactions without needing to know what order they will arrive on the network.
+
+We add an optional `random_nonce` field to the `Transaction` and `DelegateAction` messages which allows clients to pick an alternate way of disambiguating identical transactions and preventing replay attacks. The `random_nonce` is an arbitrary value selected by the client, typically a random number.
+
+We also add an optional `expires_at` field to the `Transaction` and `DelegateAction` messages. The protocol guarantees that this message will not be run in any block with a block timestamp later than the one contained in `expires_at`.
+
+`expires_at` and `random_nonce` exist as alternatives to `max_block_height` and `nonce` respectively. They will cause this field to be ignored and the mechanisms connected to them to not be used. We put limits on the validity period of messages using `random_nonce` to avoid a large cost of checking their uniqueness.
```

Comment:
Does `max_block_height` already exist on `Transaction`?  It's not in the [docs](https://nomicon.io/RuntimeSpec/Transactions).

Reply:
It [appears](https://github.com/near/nearcore/blob/master/core/primitives/src/action/delegate.rs#L36) to and it's mandatory.

Reply:
I mean for [Transaction](https://github.com/near/nearcore/blob/master/core/primitives/src/transaction.rs#L25-L40), not `DelegateAction`.  For `Transaction`, it appears that `expires_at` would be the only expiration option, not an alternative to `max_block_height`.

Reply:
I'm a bit late here but in case this is still relevant...

Today, transactions don't have an explicit `max_block_height`, as @ewiner correctly observed. Instead, they expire at the height of the given `block_hash` + `transaction_validity_period`. In mainnet, this is currently at 86400 blocks, which is 1 day if block time is 1s.

In theory, this means by picking a block hash at a specific height, one can define an expiration block height. But at the same time, the block hash is used to pick the canonical hash in case of forks. So I would argue, you really want to use the most recent `block_hash` you know, to ensure your changes build on top of the blockchain state that is known to you. But then you always get the fixed validity of 86400 blocks.

So yes, in my understanding, the proposal is the first proper way to limit a transaction validity period by the user.

Here is the check in the code:

https://github.com/near/nearcore/blob/7bad46aa5a9fe12b6d2b6e28800c7623365dd92f/chain/chain/src/store.rs#L707-L711